refactor!: update to latest brisk-reconciler

lessp opened the pull request:
~~So, pushing what I've got, I began converting the examples. Most of the should be OK now hopefully. Got a bit stuck on the `UI_Primitives` type-errors.~~

This PR introduces the changes from [brisk-reconciler](https://github.com/briskml/brisk-reconciler/commit/0a2e4765be13a47f377c836ab3eeee6cebb5d78e) with the new syntax:

e.g.

```ocaml
let%component make = () => {
  let%hook (counter, setCount) = React.Hooks.state(0);

  <Clickable onClick={_ => setCount(prevCount => prevCount + 1)}>
    <Text style=Style.[...] text={string_of_int(counter)} />
  </Clickable>
};
```

lessp commented:
Btw, this is a PR for merge into: https://github.com/revery-ui/revery/pull/566

glennsl commented on examples/DefaultButton.re:
```diff
@@ -5,7 +5,7 @@ open Revery.UI.Components;
   let%component make = () => {
     let%hook (count, setCount) = Hooks.state(0);
-    let increment = () => setCount(count + 1);
+    let increment = () => setCount(_ => count + 1);
```
You're supposed to do `setCount(count => count + 1);` here. That's why it was made a function ;) It's unlikely to cause a problem if you don't, especially in this case where it worked absolutely fine before, but when it does cause a problem it's a tricky one to find because it requires breaking the abstraction. It also might be very hard to reproduce. See briskml/brisk-reconciler#26 for details.

lessp replied:
Yep! I just mangled through all these and replaced with `_`!

lessp commented:
Ok, so with a few files commented out:

`UI_Components`
- Button.rei
- Checkbox.rei
- Clickable.rei
- ClipContainer.rei
- ClipContainer.rei
- Container.rei
- Dropdown.re
- Positioned.rei

`UI_Hooks`
- Effect.rei_
- Reducer.rei_
- Ref.rei_
- State.rei_
- Tick.rei_

`examples`
- DropdownExample.re `noop`
- FontsExample.re `noop`
- RadioButtonExample.re `noop`

-------

This builds and runs, with the core changes happening in `UI/Reconciler.re` and `UI/Container.re`

Et7f3 commented:
I don't why I have a empty windows when executing `esy run`

lessp commented:
Merged with master, some examples doesn't run:

- Fonts
- Input~
- RadioButton

Et7f3 commented:
You should better rebase onto master instead of merging (you will avoid this kind of merge error).

lessp commented:
The `Input`-component doesn't seem to be able to get refocused. I did some minor debugging and backtracked it to `NodeEvents` I think, but didn't spend a lot of time on it!

lessp commented:
The issue with the `Input`-component is fixed, turns out it was a missing default-value for `tabindex` in `Clickable` that was lost when I converted that.

I have updated all the interface-files I commented out previously. I'm still experiencing this:

> EDIT: Also, when an application starts, I've sometimes got stuck in a screen where it's not rendered completely. Not sure about that one!

Would be great if someone else could confirm!

lessp commented:
Also, running locally I'm getting the following error when trying to run tests. I think the same is happening in CI. Not sure what it means:

```
❯ esy @test run
esy-build-package: source path
                   /Users/ter/dev/diverse/reason/revery/_esy/bench/store/b/revery-cf2a28d6/install/default/lib/ReveryTest/META
                   does not exist
error: build failed with exit code: 1
  
esy: exiting due to errors above
```

Et7f3 commented on js.json:
```diff
@@ -1,29 +1,29 @@
+    "resolutions": {
+      "rebez": "github:jchavarri/rebez#46cbc183",
+      "rench": "github:revery-ui/rench#4554280",
+      "@brisk/brisk-reconciler": "github:briskml/brisk-reconciler#a0fe3a9",
```
not latest commit.

lessp replied:
Fixed in: https://github.com/revery-ui/revery/pull/576/commits/1db02312dfa80c49b01bffd907340ebf81f091b2

Et7f3 commented on src/UI/Container.re:
```diff
@@ -1,21 +1,30 @@
+              deleteNode: React.deleteNode,
+              moveNode: React.moveNode,
+            },
+            React.empty,
```
Where is element ? is it normal it is React.empty ?

lessp replied:
Ahh, nice catch! This fixed the rendering issue where it would not render the content before you had given the window focus!

lessp replied:
Fixed in: https://github.com/revery-ui/revery/pull/576/commits/4c68021239363ea3fe8da0e344a34ff9ff0bfa2d

Et7f3 commented on src/UI_Components/Button.rei:
```diff
@@ -11,10 +11,9 @@ Simple out-of-box button component
     ~onClick: unit => unit=?,
-    ~color: Revery_Core.Color.t=?,
+    ~color: Revery_Core__.Color.t=?,
     ~fontSize: int=?,
```
Why ? and String.t

lessp replied:
Perfect, thanks!

lessp replied:
Fixed in: https://github.com/revery-ui/revery/pull/576/commits/9516740a91cefdd4a55a0cbdfd7e33e221c39498

Et7f3 commented on src/UI_Components/Clickable.re:
```diff
@@ -23,34 +23,31 @@ let isMouseInsideRef = (ref: node, mouseX: float, mouseY: float) => {
+                ~componentRef=?,
+                ~onBlur=?,
+                ~onFocus=?,
+                ~tabindex=0,
```
why 0? Many node will have zero as index...

lessp replied:
Good question, I'm not sure, but my feeling was that this is sometimes needed for an element to be able to receive focus. 

I had forgotten it initially, when added back it allowed the `Input`-component to again receive focus. 

Anyway, the default-value here is not a part of this commit, so might be a different discussion!

<https://github.com/revery-ui/revery/blob/master/src/UI_Components/Clickable.re#L114>

Et7f3 replied:
I have applied diff from HEAD of master 4a37510cc8ca436e69fc8e77e9930126ca633b33 and this branch.

Et7f3 commented on src/UI_Components/ExpandContainer.re:
```diff
@@ -2,10 +2,9 @@ open Revery_Core;
 
-[@component]
-let make = (~children, ~color=Colors.transparentWhite, (), hooks) => {
+let make = (~children, ~color=Colors.transparentWhite, ()) => {
```
shouldn't be `let%component make` or it is stateless so no need for this extention ?

lessp replied:
Exactly, it's stateless, so can't use the extension here.

Et7f3 commented on src/UI_Hooks/Animation.rei:
```diff
@@ -1,13 +1,62 @@
+        (
+          ref(Revery_UI.Animated.animation),
+          Revery_UI.React.Hooks.Reducer.t(int),
+          Revery_UI.React.Hooks.Effect.t(Revery_UI.React.Hooks.Effect.onMount)
```
Why so long name 😂 `Reducer.t(int)` instead of `Revery_UI.React.Hooks.Reducer.t(int)`

Et7f3 replied:
Fixed in 9516740

Et7f3 commented on test/UI/ReconcilerTests.re:
```diff
@@ -1,41 +1,39 @@
 
 module TestRefComponent = {
-  let component = React.component("TestRefComponent");
+  [@component]
```
Old syntax.

lessp replied:
Fixed in: https://github.com/revery-ui/revery/pull/576/commits/245f46f8b651b227c43973700a5f4992ad55ba67

Et7f3 commented on src/UI_Primitives/Image.re:
```diff
@@ -17,25 +17,23 @@ let getStyles: (option(int), option(int), Style.t) => Style.t =
+                      ~height=?,
+                      ~src="",
+                      ~style=Style.emptyImageStyle,
+                      ~children=React.empty,
```
Image can have children 👀

Et7f3 replied:
Same for other Primitives like OpenGL

lessp replied:
I don't think this changes that, we just need t provide an empty children as a default which is needed IIRC!

Et7f3 replied:
Scrollview, Row, Container, Center must have one argument so default arg to `React.empty` is not fine for me. Ticker shouldn't have child.

Et7f3 replied:
I though not exposing argument and defining a local children variable. What do you think ? I can fix that.

lessp replied:
That might make sense, but it differs from how master looks currently: <https://github.com/revery-ui/revery/blob/master/src/UI_Components/Input.re#L356>

lessp replied:
I've updated all containers except `Container` as to explicitly take children @Et7f3. 
Do you want me to push these, or do you want to create a separate issue for that?

Et7f3 replied:
Ah I see so create another issue. IIRC @bryphe wanted some compile time assertion to be verified.

Et7f3 replied:
for reference https://github.com/revery-ui/revery/issues/489#issuecomment-546668675_

lessp commented:
Wondering if we should change the merge target to master instead of #566?

Et7f3 commented:
When reviewing it I compared this branch and master and it was clean. I change it.

bryphe commented:
So cool to see this green and the `let%hooks` land! 🎉 Thanks for all the work / help with this!

> Wonder if we should change the merge target to master instead of #566?

Sounds good to me! We can just close #566 in lieu of this more complete change 😄

wokalski commented:
The only problem is that let%hook doesn't work in reis currently. And it's hard to make it work there.

lessp commented:
> The only problem is that let%hook doesn't work in reis currently. And it's hard to make it work there.

Would you like to get this in before merging, or should we add it at a later stage? @wokalski @Et7f3

wokalski commented:
No, it’s very hard. There are lower hanging fruits with more ROI.

lessp commented on src/UI_Components/Positioned.re:
```diff
@@ -1,22 +1,21 @@
 open Revery_UI_Primitives;
 
-let component = React.component("Positioned");
+module X = Revery_UI.Brisk_jsx_runtime;
```
I wouldn't think so! I removed them. I wonder if they were used in the initial refactor in https://github.com/revery-ui/revery/pull/566. Anyway, good catch!

Removed in: 56e6c33 and a47c9dc

wokalski replied:
Brisk_jsx_runtime is unnecessary as of the defunctorized brisk-reconciler!

bryphe commented:
Awesome! Let's bring this in - very exciting! 🎉 Thanks @wokalski @lessp @Et7f3
